### PR TITLE
Resize

### PR DESCRIPTION
**Pull request by AlexeyKozhevin**
Add separable transpose conv, bilinear resize and depth to space transformations for dim = 1,2,3
Add conv1d_transpose in tf.nn style
Refactor conv_block.py. Now there are conv_block.py (main file), block.py and resize.py
Rename 's' to 'C' in layout and add 'T' for separable transpose conv

### CODE REVIEW

File: dataset/models/tf/layers/conv.py [hunk 1]
```diff
@@ -26,7 +17,13 @@ def conv(inputs, *args, **kwargs):
...
     """
     dim = inputs.shape.ndims - 2
-    layer_fn = _CONV_LAYERS[dim]
+    conv_layers = {
+        1: tf.layers.conv1d,
+        2: tf.layers.conv2d,
+        3: tf.layers.conv3d
+    }
+    layer_fn = conv_layers[dim]
+    print(args, kwargs)
```
Comment (roman-kh) [inline dataset/models/tf/layers/conv.py]: print

File: dataset/models/tf/layers/resize.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    data_format : {'channels_last', 'channels_first'}
+        position of the channels dimension
+
+    Returns
+    -------
+    tf.Tensor
+    """
+    dim = inputs.shape.ndims
+    if dim != 4:
+        raise ValueError("inputs must be Tensor of rank but {} was given".format(dim))
```
Comment (roman-kh) [inline dataset/models/tf/layers/resize.py]: of which rank?

File: dataset/models/tf/layers/block.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+
+    if layout_dict is not None:
+        nd_layers = {**nd_layers, **layout_dict[0]}
+        c_layers = {**c_layers, **layout_dict[1]}
+    _layers_keys = str(list(c_layers.keys()))
+    _group_keys = (
+        _layers_keys
+        .replace('t', 'c')
+        .replace('s', 'c')
```
Comment (roman-kh) [inline dataset/models/tf/layers/block.py]: remove `s`, add `T` and `C`

File: dataset/models/tf/layers/block.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            else:
+                arg_value = args[arg]
+            new_args.update({arg: arg_value})
+
+        return new_args
+
+    layout = layout or ''
+    layout = layout.replace(' ', '')
+    if len(layout) == 0:
+        logging.warning('conv_block: layout is empty, so there is nothing to do, just returning inputs.')
```
Comment (roman-kh) [inline dataset/models/tf/layers/block.py]: [create logger first](https://docs.python.org/3/library/logging.html#logging.getLogger)

File: dataset/models/tf/layers/conv.py [hunk 5]
```diff
@@ -26,7 +17,13 @@ def conv(inputs, *args, **kwargs):
     `tf.layers.conv3d <https://www.tensorflow.org/api_docs/python/tf/layers/conv3d>`_
     """
     dim = inputs.shape.ndims - 2
-    layer_fn = _CONV_LAYERS[dim]
+    conv_layers = {
+        1: tf.layers.conv1d,
+        2: tf.layers.conv2d,
+        3: tf.layers.conv3d
+    }
```
Comment (roman-kh) [inline dataset/models/tf/layers/conv.py]: There is no point in creating this dict in every call. Create it once when the module is initialized or don't create it at all (use `if` instead)

File: dataset/models/tf/layers/conv.py [hunk 6]
```diff
@@ -96,8 +135,71 @@ def conv_transpose(inputs, filters, kernel_size, strides, *args, **kwargs):
     return output
 
 
+def _common_separable_conv(transpose, inputs, filters, kernel_size, strides, padding, data_format,
```
Comment (roman-kh) [inline dataset/models/tf/layers/conv.py]: _separable_conv

File: dataset/models/tf/layers/conv.py [hunk 7]
```diff
@@ -96,8 +134,71 @@ def conv_transpose(inputs, filters, kernel_size, strides, *args, **kwargs):
     return output
 
 
+def _separable_conv(transpose, inputs, filters, kernel_size, strides, padding, data_format,
```
Comment (roman-kh) [inline dataset/models/tf/layers/conv.py]: Add defaults starting from `strides` or even `kernel_size`

File: dataset/models/tf/layers/conv.py [hunk 8]
```diff
@@ -96,8 +134,71 @@ def conv_transpose(inputs, filters, kernel_size, strides, *args, **kwargs):
...
+    kwargs = {'kernel_size': kernel_size,
+              'strides': strides,
+              'padding': padding,
+              'data_format': data_format,
+              'dilation_rate': dilation_rate,
+              'activation': activation,
+              **kwargs}
+
+    if transpose:
+        kwargs.pop('dilation_rate')
```
Comment (roman-kh) [inline dataset/models/tf/layers/conv.py]: Add `dilation_rate` only when needed (not add it always and then remove sometimes)

File: dataset/models/tf/layers/conv.py [hunk 9]
```diff
@@ -96,8 +134,71 @@ def conv_transpose(inputs, filters, kernel_size, strides, *args, **kwargs):
...
+
+    if channels_in * depth_multiplier != filters:
+        _kwargs = {**kwargs,
+                   'inputs': depthwise_conv,
+                   'filters': filters,
+                   'kernel_size': 1,
+                   'strides': 1,
+                   'name': 'pointwise'}
+        if not transpose:
+            kwargs['dilation_rate'] = 1
```
Comment (roman-kh) [inline dataset/models/tf/layers/conv.py]: _kwargs?
As 1 is a default values may be it is not needed at all.

File: dataset/models/tf/layers/conv_block.py [hunk 10]
```diff
@@ -206,120 +148,10 @@ def conv_block(inputs, layout='', filters=0, kernel_size=3, name=None,
...
-                with tf.variable_scope('layer-%d' % i):
-                    tensor = layer_fn(tensor, **args)
-
-    if context is not None:
-        context.__exit__(None, None, None)
-
+    tensor = _conv_block(inputs, layout, filters, kernel_size, name,
+                         strides, padding, data_format, dilation_rate, depth_multiplier,
+                         activation, pool_size, pool_strides, dropout_rate, is_training,
+                         [ND_LAYERS, C_LAYERS], **kwargs)
```
Comment (roman-kh) [inline dataset/models/tf/layers/conv_block.py]: make layer dicts keyword arguments to prevent future collisions.

File: dataset/models/tf/layers/resize.py [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+    return shape
+
+
+def depth_to_space(inputs, block_size, name='d2s', data_format='channels_last'):
+    """ 2d and 3d depth_to_space transformation.
+
+    Parameters
+    ----------
+    inputs : tf.Tensor
+
```
Comment (roman-kh) [inline dataset/models/tf/layers/resize.py]: add some text

File: dataset/models/tf/layers/resize.py [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+    block_size : int
+
```
Comment (roman-kh) [inline dataset/models/tf/layers/resize.py]: add some text

File: dataset/models/tf/layers/resize.py [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+    name : str
+        scope name
+
```
Comment (roman-kh) [inline dataset/models/tf/layers/resize.py]: remove empty line

File: dataset/models/tf/layers/conv.py (hunk 9 above)
Comment (AlexeyKozhevin) [reply dataset/models/tf/layers/conv.py]: If dilation rate for previous conv layers was not 1 then we must rewrite it.

File: dataset/models/tf/gcn.py [hunk 14]
```diff
@@ -41,7 +41,8 @@ def default_config(cls):
         config = TFModel.default_config()
 
         config['input_block'].update(dict(layout='cna', filters=64, kernel_size=7, strides=2))
-        config['body']['encoder'] = dict(base_class=ResNet101, filters=[256, 512, 1024, 2048])
+        #config['body']['encoder'] = dict(base_class=ResNet101, filters=[256, 512, 1024, 2048])
+        config['body']['encoder'] = dict(base_class=ResNet101, filters=[64, 128, 256, 512])
```
Comment (roman-kh) [inline dataset/models/tf/gcn.py]: Exclude `gcn` from pull request
